crates/ruff_linter/src/rules/flake8_comprehensions/rules/unnecessary_comprehension.rs

```diff
@@ -30,10 +29,25 @@ use crate::rules::flake8_comprehensions::fixes;
 /// set(iterable)
 /// ```
 ///
+/// ## Known problems
+///
+/// When creating a dictionary by iterating over a type that yields different values during
+/// `__iter__` than `keys()` and `__getitem__`, this rule may produce false positives. For example:
+///
+/// ```python
+/// d1 = {(1, 2): 3, (4, 5): 6}
+/// d2 = {x: y for x, y in d1}
+/// ```
+///
+/// Iterating over `d1` yields the keys to the dictionary. However, Ruff will suggest using
+/// `d2 = dict(d1)` instead which constructs a new dictionary by iterating over the keys _and_
+/// retrieving the associated items. Ruff cannot consistently infer if a type will be valid in this
+/// context.
```

Comment:
How about something like this?

```suggestion
/// This rule may produce false positives for `dict` comprehensions due to the fact that
/// the `dict` constructor method behaves differently depending on whether it is passed
/// a sequence (such as a `list`) or a mapping (such as a `dict`). If a `dict` comprehension
/// iterates over the keys of an existing dictionary to create a new one, rewriting the
/// comprehension to use `dict()` directly according to this rule's suggestion will not
/// produce the same result.
///
/// For example:
/// ```pycon
/// >>> d1 = {(1, 2): 3, (4, 5): 6}
/// >>> {x: y for x, y in d1}  # iterates over the keys and unpacks each tuple key
/// {1: 2, 4: 5}
/// >>> dict(d1)               # creates an exact copy of the original mapping
/// {(1, 2): 3, (4, 5): 6}
/// ```
///
/// The suggested change is still valid for `dict` comprehensions if the object being
/// iterated over is a `list` or some other iterable, however. Ruff currently has
/// limited type inference, so cannot reliably infer if the type of the object being
/// iterated over makes the suggestion valid.
```

Reply:
I took parts of this, but edited it for consistency with #13629 and clarity.